File: app/src/main/java/com/shalzz/attendance/data/local/DatabaseHelper.java
package com.shalzz.attendance.data.local;

import android.database.sqlite.SQLiteDatabase;

import com.shalzz.attendance.data.model.AbsentDate;
import com.shalzz.attendance.data.model.ListFooter;
import com.shalzz.attendance.data.model.Period;
import com.shalzz.attendance.data.model.Subject;
import com.shalzz.attendance.data.model.User;
import com.shalzz.attendance.model.SubjectModel;
import com.shalzz.attendance.wrapper.DateHelper;
import com.squareup.sqlbrite.BriteDatabase;
import com.squareup.sqlbrite.SqlBrite;
import com.squareup.sqldelight.SqlDelightStatement;

import java.util.Collection;
import java.util.Collections;
import java.util.Date;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;
import rx.schedulers.Schedulers;

/**
 * Helper Class for SQLite database
 *
 * @author shalzz
 */
@Singleton
public class DatabaseHelper {

    private final BriteDatabase mDb;
    private final Subject.InsertSubject insertSubject;

    @Inject
    public DatabaseHelper(DbOpenHelper dbOpenHelper) {
        SqlBrite sqlBrite = new SqlBrite.Builder().build();
        mDb = sqlBrite.wrapDatabaseHelper(dbOpenHelper, Schedulers.io());
//        mDb.setLoggingEnabled(true);

        SQLiteDatabase db = dbOpenHelper.getWritableDatabase();
        insertSubject = new SubjectModel.InsertSubject(db);
    }

    public BriteDatabase getBriteDb() {
        return mDb;
    }

    public Observable<Subject> setSubjects(final Collection<Subject> newSubjects) {
        return Observable.create(subscriber -> {
            if (subscriber.isUnsubscribed()) return;

            try (BriteDatabase.Transaction transaction = mDb.newTransaction()) {
                mDb.delete(Subject.TABLE_NAME, null);
                for (Subject subject : newSubjects) {
                    insertSubject.bind(subject.id(),
                            subject.name(),
                            subject.attended(),
                            subject.held());
                    long result = mDb.executeInsert(Subject.TABLE_NAME, insertSubject.program);

                    // Store the dates in another table corresponding to the same id
                    if (subject.absent_dates() != null) {
                        for (Date date : subject.absent_dates()) {
                            mDb.insert(AbsentDate.TABLE_NAME,
                                    AbsentDate.FACTORY.marshal()
                                            .subject_id(subject.id())
                                            .absent_date(date)
                                            .asContentValues());
                        }
                    }
                    if (result >= 0) subscriber.onNext(subject);
                }
                transaction.markSuccessful();
                subscriber.onCompleted();
            }
        });
    }

    public Observable<List<Subject>> getSubjects(String filter) {
        filter = filter == null ? "" : filter;
        filter = '%' + filter + '%';
        SqlDelightStatement query = Subject.FACTORY.selectLikeName(filter);
        return mDb.createQuery(query.tables, query.statement, query.args)
                .mapToList(Subject.MAPPER::map);
    }

    /**
     * Gets a list of subjects that are marked absent on a particular date
     *
     * @param date The date
     * @return A list of Id's of subjects marked as absent
     */
    public Observable<List<Integer>> getAbsentSubjects(Date date) {
        SqlDelightStatement query = AbsentDate.FACTORY.select_absent_subjects(date);
        return mDb.createQuery(query.tables, query.statement, query.args)
                .mapToList(cursor -> AbsentDate.MAPPER.map(cursor).subject_id());
    }

    public Observable<Period> addPeriods(final List<Period> newPeriods) {
        return Observable.create(subscriber -> {
            if (subscriber.isUnsubscribed()) return;
            if (newPeriods.isEmpty()) return;

            try (BriteDatabase.Transaction transaction = mDb.newTransaction()) {
                mDb.delete(Period.TABLE_NAME, "date = ?", newPeriods.get(0).date());
                for (Period period : newPeriods) {
                    long result = mDb.insert(Period.TABLE_NAME,
                            Period.FACTORY.marshal(period).asContentValues(),
                            SQLiteDatabase.CONFLICT_REPLACE);
                    if (result >= 0) subscriber.onNext(period);
                }
                transaction.markSuccessful();
                subscriber.onCompleted();
            }
        });
    }

    public Observable<List<Period>> getPeriods(Date date) {
        SqlDelightStatement query = Period.FACTORY.select_by_date(DateHelper.formatToTechnicalFormat(date));
        return mDb.createQuery(query.tables, query.statement, query.args)
                .mapToList(Period.MAPPER::map);
    }

    public Observable<User> addUser(User user) {
        return Observable.create(subscriber -> {
            if (subscriber.isUnsubscribed()) return;
            long result = mDb.insert(User.TABLE_NAME, User.FACTORY.marshal(user).asContentValues(),
                    SQLiteDatabase.CONFLICT_REPLACE);
            if (result >= 0) subscriber.onNext(user);
            subscriber.onCompleted();
        });
    }

    public Observable<User> getUser() {
        SqlDelightStatement query = User.FACTORY.select_all();
        return mDb.createQuery(query.tables, query.statement, query.args)
                .mapToOne(User.MAPPER::map);
    }

    public Observable<ListFooter> getListFooter() {
        SqlDelightStatement query = Subject.FACTORY.selectTotal();
        return mDb.createQuery(query.tables, query.statement, query.args)
                .mapToOne(cursor -> ListFooter.builder()
                        .setAttended(cursor.getFloat(0))
                        .setHeld(cursor.getFloat(1))
                        .build());
    }

    public Observable<Integer> getSubjectCount() {
        SqlDelightStatement query = Subject.FACTORY.selectCount();
        return mDb.createQuery(query.tables, query.statement, query.args)
                .mapToOne(cursor -> cursor.getInt(0));
    }

    public Observable<Integer> getPeriodCount(Date day) {
        SqlDelightStatement query =
                Period.FACTORY.select_count_by_date(DateHelper.formatToTechnicalFormat(day));
        return mDb.createQuery(query.tables, query.statement, query.args)
                .mapToOne(cursor -> cursor.getInt(0));
    }

    public Observable<Integer> getUserCount() {
        SqlDelightStatement query = User.FACTORY.select_count();
        return mDb.createQuery(query.tables, query.statement, query.args)
                .mapToOne(cursor -> cursor.getInt(0));
    }

    /**
     * Delete All Rows
     * */
    public void resetTables(){
        mDb.delete(Subject.TABLE_NAME, null);
        mDb.delete(Period.TABLE_NAME, null);
        mDb.delete(User.TABLE_NAME, null);
        mDb.delete(AbsentDate.TABLE_NAME, null);
    }
}

File: app/src/main/java/com/shalzz/attendance/data/local/DatabaseHelper.kt
/*
 * Copyright (c) 2013-2018 Shaleen Jain <shaleen.jain95@gmail.com>
 *
 * This file is part of College Academics.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.shalzz.attendance.data.local

import com.shalzz.attendance.data.model.ListFooter
import com.shalzz.attendance.data.model.entity.Period
import com.shalzz.attendance.data.model.entity.Subject
import com.shalzz.attendance.data.model.entity.User
import com.shalzz.attendance.wrapper.DateHelper
import io.reactivex.Observable
import io.reactivex.Single
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import java.util.Date
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Helper Class for SQLite database
 *
 * @author shalzz
 */
@Singleton
class DatabaseHelper @Inject
constructor(private val mDb: AppDatabase) {

    val listFooter: Observable<ListFooter>
        get() = mDb.subjectDao().getTotalAttendance()

    val subjectCount: Single<Int>
        get() = mDb.subjectDao().getCount()

    val userCount: Single<Int>
        get() = mDb.userDao().getCount()

    fun getUser(username: String): Observable<User> {
        return mDb.userDao().getAllByUsername(username)
    }

    fun setSubjects(newSubjects: Collection<Subject>): Observable<Subject> {
        return Observable.create { source ->
            if (source.isDisposed) return@create

            mDb.runInTransaction {
                mDb.subjectDao().deleteAll()
                for (subject in newSubjects) {
                    mDb.subjectDao().insert(subject)
                    source.onNext(subject)
                }
                source.onComplete()
            }
        }
    }

    fun getSubjects(filter: String?): Observable<List<Subject>> {
        val name = if (filter == null) "%%" else "%$filter%"
        return mDb.subjectDao().getAllLikeName(name)
    }

    fun setPeriods(newPeriods: List<Period>): Observable<Period> {
        return Observable.create { subscriber ->
            if (subscriber.isDisposed) return@create
            if (newPeriods.isEmpty()) return@create

            mDb.runInTransaction {
                mDb.periodDao().deleteByDate(newPeriods[0].date)
                for (period in newPeriods) {
                    mDb.periodDao().insert(period)
                    subscriber.onNext(period)
                }
                subscriber.onComplete()
            }
        }
    }

    fun getPeriods(date: Date): Observable<List<Period>> {
        return mDb.periodDao().getAllByDate(DateHelper.toTechnicalFormat(date))
    }

    fun setUser(user: User): Observable<User> {
        return Observable.create { subscriber ->
            if (subscriber.isDisposed) return@create
            mDb.userDao().insert(user)
            subscriber.onNext(user)
            subscriber.onComplete()
        }
    }

    fun getPeriodCount(day: Date): Single<Int> {
        return mDb.periodDao().getCountByDate(DateHelper.toTechnicalFormat(day))
    }

    /**
     * Delete All Rows
     */
    fun resetTables() : Job  {
        return GlobalScope.launch (Dispatchers.IO) {
            mDb.subjectDao().deleteAll()
            mDb.periodDao().deleteAll()
            mDb.userDao().deleteAll()
        }
    }
}
